### PR TITLE
zkVM: split binfmt feature to a new "risc0-binfmt" crate for modularity

### PR DESCRIPTION
**Pull request by SchmErik**
This PR splits off the `binfmt` feature to a separate crate in order to avoid a circular dependency with a zkVM capability to be added in the near future. This capability requires risc0-zkvm to be dependent on risc0-build and the zkVM crate will contain guest code used in cases where the executor runs into a fault.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -33,6 +34,7 @@ homepage = "https://risczero.com/"
 repository = "https://github.com/risc0/risc0/"
 
 [workspace.dependencies]
+risc0-binfmt         = { version = "0.16.1", default-features = false, path = "risc0/binfmt" }
```
Comment (flaub) [inline Cargo.toml:37]: Could you sort this?

File: examples/digital-signature/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,8 @@ sha2 = "0.10"
 
 [dev-dependencies]
 test-log = "0.2"
+tracing = { version = "0.1", default-features = false, features = ["attributes"] }
+tracing-subscriber = { version = "0.3", features = ["env-filter"] }
```
Comment (flaub) [inline examples/digital-signature/Cargo.toml:19]: Is this related?

File: risc0/binfmt/src/elf.rs [hunk 3]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+extern crate alloc;
```
Comment (flaub) [inline risc0/binfmt/src/elf.rs:15]: I think this should only go in `lib.rs`

Comment (flaub) [reply risc0/binfmt/src/elf.rs:15]: Actually, since we require `std`, maybe just use `std::collections::BTreeMap`

File: examples/digital-signature/Cargo.toml (hunk 2 above)
Comment (SchmErik) [reply examples/digital-signature/Cargo.toml]: Yes, So... oddly enough. This PR made this example fail on `cargo test` with the following error and adding this helped the test pass. I'm not sure what exactly changed but it fails to compile without these additional deps now.
```
% cargo test
   Compiling risc0-binfmt v0.16.1 (/Users/erik/risc0/risc0/binfmt)
   Compiling risc0-build v0.16.1 (/Users/erik/risc0/risc0/build)
   Compiling risc0-zkvm v0.16.1 (/Users/erik/risc0/risc0/zkvm)
   Compiling digital-signature-methods v0.1.0 (/Users/erik/risc0/examples/digital-signature/methods)
   Compiling digital-signature-core v0.1.0 (/Users/erik/risc0/examples/digital-signature/core)
sign: Starting build for riscv32im-risc0-zkvm-elf digital-signature-methods(build), risc0-zkvm                                                                           
sign:     Updating crates.io index
sign:     Finished release [optimized] target(s) in 1.37s-signature-methods(build)                                                                                       
   Compiling digital-signature v0.1.0 (/Users/erik/risc0/examples/digital-signature)
error[E0433]: failed to resolve: could not find `tracing_subscriber` in the list of imported crates
  --> digital-signature/src/lib.rs:72:5
   |
72 |     #[test]
   |     ^^^^^^^ could not find `tracing_subscriber` in the list of imported crates
   |
   = note: this error originates in the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `tracing` in the list of imported crates
  --> digital-signature/src/lib.rs:72:5
   |
72 |     #[test]
   |     ^^^^^^^ could not find `tracing` in the list of imported crates
   |
   = note: this error originates in the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0433`.
error: could not compile `digital-signature` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

**Review by flaub — APPROVED** [review]

File: risc0/binfmt/src/elf.rs [hunk 4]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use alloc::collections::BTreeMap;
```
Comment (SkymanOne) [inline risc0/binfmt/src/elf.rs:15]: This makes `risc0_zkvm` and `risc0_binfmt` to be `no_std` incompatible
